dcpy/models/product/dataset/metadata_v2.py

```diff
@@ -0,0 +1,424 @@
+
+    Note: This is put in its own class because it might be useful for other
+    classes unrelated to product metadata in the future.
+    """
+
+    _exclude_falsey_values: bool = True
+    _head_sort_order: list[str] = PrivateAttr(default=["id"])
+    _tail_sort_order: list[str] = PrivateAttr(default=["custom"])
+
+    @model_serializer(mode="wrap")
+    def _model_dump_ordered(self, handler):
+        unordered = handler(self)
+
+        ordered_items_head = []
+        ordered_items_tail = []
+        simple_type_items = []
+        other_items = []
+
+        for model_field in list(unordered.items()):
+            if not model_field[1] and self._exclude_falsey_values:
+                # If an object's values are all None, it will serialize as {}.
+                # These aren't removed by model_dump(exclude_none=True), so we have to do it manually.
+                continue
+            field_type = self.model_fields[
+                model_field[0]
+            ].annotation  # Need to retrieve type from the class def, not the instance
+            is_literal = type(field_type) is typing._LiteralGenericAlias  # type: ignore
+            simple_types = {
+                bool,
+                bool | None,  # This is a little hacky, but does the job.
+                str,
+                str | None,
+                int,
+                int | None,
+                float,
+                float | None,
+                type(None),
+            }
+
+            if model_field[0] in self._head_sort_order:
+                ordered_items_head.append(model_field)
+            elif model_field[0] in self._tail_sort_order:
+                ordered_items_tail.append(model_field)
+            elif field_type in simple_types or is_literal:
+                simple_type_items.append(model_field)
+            else:
+                other_items.append(model_field)
+
+        # As of python 3.7, dict keys are ordered, and so will serialize in the order below
+        return dict(
+            sorted(ordered_items_head, key=lambda x: self._head_sort_order.index(x[0]))
+            + sorted(simple_type_items, key=lambda x: x[0])
+            + sorted(other_items, key=lambda x: x[0])
+            + sorted(
+                ordered_items_tail, key=lambda x: self._tail_sort_order.index(x[0])
+            )
+        )
+
+
+class CustomizableBase(SortedSerializedBase, extra="forbid"):
+    """A Base Pydantic class to allow extensibility of our models via a `custom`
+    dictionary.
+
+    Any additional attributes that aren't defined on our models should
+    be added to `custom`. This is intended for domain-specific, non-generalized uses
+    like data-dictionary generation (e.g. the `readme_data_type` field on the columns)
+
+    It's also important that custom be preserved in the
+    course of overriding, specifically with a deep merge of the dictionary elements.
+    """
+
+    custom: dict[str, Any] = {}
+
+
+class YamlWriter(BaseModel):
```

Comment:
It would be nice to move this to "util" models or something like that as it doesn't feel to be specific to metadata model exclusively

Reply:
def agree. Moved [here](https://github.com/NYCPlanning/data-engineering/pull/1069/commits/62f2acb68d82645249223110a8e80827aef599ca) to a file called `models/base.py` - thinking we'll eventually have a few top-level base models we'd like to mix in. Open to other names though!